### PR TITLE
New HvH gamemode Sensor Capture [MDB ignore] pt2

### PR DESCRIPTION
**Pull request by tyeagg**
## About The Pull Request
this is basically combat patrol but crash but on a faster pace kinda.
Marines must capture and activate 5 sensor towers and hold it for 150 seconds, activating sensor towers will increase game end timer and if all towers are fully activated marines win.
SOM must stop marines from reactivating all towers until the game end timer is over

## Why It's Good For The Game
could be fun

## Changelog
:cl:
add: sensor capture gamemode
imageadd: sensor tower minimap blips
/:cl:

### CODE REVIEW

**Comment by MereWalker** [comment]
Be sure there are a few places only som can get to and out of so marines can't spawn camp them. Just a thought. Look forward to having an additional hvh mode! Best of luck and can't wait for more updates!

**Comment by Autisem** [comment]
Giving objectives would imo mape CP much better already.

**Comment by Lumipharon** [comment]
Unless I'm misunderstanding, all 5 towers have to be activated, and you have to stop the SOM from deactivating them?

That sounds quite difficult to manage as splitting your forces is usually death, while SOM can send one funni man to try recap towers.

**Comment by tyeagg** [comment]
> Unless I'm misunderstanding, all 5 towers have to be activated, and you have to stop the SOM from deactivating them?
> 
> That sounds quite difficult to manage as splitting your forces is usually death, while SOM can send one funni man to try recap towers.

**Comment by Lumipharon** [comment]
> no you just have to hold a tower for 150 seconds and it adds to a point pool and than som cant deactivate it

Ah cool, that makes more sense. It does put the shoe on the other foot a bit though, in that SOM will need to be able to respond to marines capping multiple towers, although that's probably less of an issue.

File: code/datums/gamemodes/game_mode.dm [hunk 1]
```diff
@@ -26,6 +26,8 @@
 	var/win_points_needed = 0
 	///The points per faction, assoc list
 	var/list/points_per_faction
+	///The amount of activated sensor towers in sensor capture
+	var/sensors_activated = 0
```
Comment (TiviPlus) [inline code/datums/gamemodes/game_mode.dm:30]: This should be a sensor specific var

File: code/game/objects/effects/landmarks/landmarks.dm [hunk 2]
```diff
@@ -430,6 +430,16 @@
 	GLOB.sensor_towers += loc
 	return INITIALIZE_HINT_QDEL
 
+/obj/effect/landmark/sensor_tower_patrol
+	name = "Sensor tower"
+	icon = 'icons/obj/structures/sensor.dmi'
+	icon_state = "sensor_loyalist"
+
+/obj/effect/landmark/sensor_tower_patrol/Initialize()
+	. = ..()
```
Comment (TiviPlus) [inline code/game/objects/effects/landmarks/landmarks.dm:439]:
```suggestion
	..()
```

File: code/game/objects/structures/sensor_tower.dm [hunk 3]
```diff
@@ -61,3 +61,101 @@
...
+	density = TRUE
+	layer = BELOW_OBJ_LAYER
+	resistance_flags = RESIST_ALL
+	var/current_timer
+	var/generate_time = 150 SECONDS
+	var/activate_time = 5 SECONDS // time to start the activation
+	var/deactivate_time = 10 SECONDS // time to stop the activation proccess
+	var/static/id = 1
+	var/towerid
+	var/activated = FALSE // if all segments are finished
```
Comment (TiviPlus) [inline code/game/objects/structures/sensor_tower.dm:79]: This entire object and it's procs are missing autodoc

File: code/game/objects/structures/sensor_tower.dm [hunk 4]
```diff
@@ -61,3 +61,101 @@
...
+	if(activated)
+		balloon_alert(user, "[src] is already fully activated!")
+		return
+	if(current_timer)
+		balloon_alert(user, "[src] is currently activating!")
+		return
+	balloon_alert_to_viewers("[user] starts to activate [src]!")
+	if(!do_after(user, activate_time, TRUE, src))
+		return
+	balloon_alert_to_viewers("[user] activates [src]!")
```
Comment (TiviPlus) [inline code/game/objects/structures/sensor_tower.dm:120]: dont use [src] in baloon alerts and even generally avoid inserting names
these should be as short as possible, and remember that players can see where it is on the map (so no specified what is being affected needed)

File: code/game/objects/structures/sensor_tower.dm [hunk 5]
```diff
@@ -61,3 +61,101 @@
...
+		return
+	playsound(src, 'sound/machines/ping.ogg', 25, 1)
+	current_timer = null
+	balloon_alert_to_viewers("[src] has finished activation!")
+	for(var/mob/living/carbon/human/human AS in GLOB.alive_human_list)
+		human.playsound_local(human, "sound/effects/CIC_order.ogg", 10, 1)
+		human.play_screen_text("<span class='maptext' style=font-size:24pt;text-align:center valign='top'><u>OVERWATCH</u></span><br>" + "[src] is fully activated.", /obj/screen/text/screen_text/command_order)
+	SSticker.mode.sensors_activated += 1
+	var/datum/game_mode/combat_patrol/sensor_capture/D = SSticker.mode
+	var/current_time = timeleft(D.game_timer)
```
Comment (TiviPlus) [inline code/game/objects/structures/sensor_tower.dm:141]: Not going to force you but generally single name vars are not cool

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/sensor_tower.dm [hunk 6]
```diff
@@ -61,3 +61,113 @@
...
+	for(var/mob/living/carbon/human/human AS in GLOB.alive_human_list)
+		human.playsound_local(human, "sound/effects/CIC_order.ogg", 10, 1)
+		human.play_screen_text("<span class='maptext' style=font-size:24pt;text-align:center valign='top'><u>OVERWATCH</u></span><br>" + "[src] activation process has been stopped.", /obj/screen/text/screen_text/command_order)
+	update_icon()
+
+/obj/structure/sensor_tower_patrol/update_icon()
+	. = ..()
+	update_control_minimap_icon()
+
+/obj/structure/sensor_tower_patrol/proc/update_control_minimap_icon()
```
Comment (TiviPlus) [inline code/game/objects/structures/sensor_tower.dm:169]: autodoc

File: code/game/objects/effects/landmarks/landmarks.dm (hunk 2 above)
Comment (TiviPlus) [reply code/game/objects/effects/landmarks/landmarks.dm:439]: Request still pending

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Rest LGTM

**Pull request merged**